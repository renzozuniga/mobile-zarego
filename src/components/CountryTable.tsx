import React, { useEffect, useState, useRef } from 'react';
import { View, Text, StyleSheet, ScrollView } from 'react-native';
import Axios from "axios";
import { REACT_MOBILE_BACKEND_ZAREGO } from "@env";
import { Table, TableWrapper, Row } from 'react-native-table-component';

const initialPage = `${REACT_MOBILE_BACKEND_ZAREGO}/api/countries?page=1&rows=100`;

const borderColor = '#C1C0B9';
const primaryColor = 'black';

const CountryTable = ({ route }: any) => {
    const { list } = route.params
    const [data, setData] = useState([]);

    const leftRef = useRef<ScrollView>(null);
    const rightRef = useRef<ScrollView>(null);
    const [tableHead, setTableHead] = useState([]);
    const [widthArr, setWidthArr] = useState([50, 60, 80, 100, 120, 140, 160, 180, 200]);
    const [recData, setRecData] = useState<any>([]);
    const [tabData, setTabData] = useState<any>([]);

    const headerHeight = 40;
    const leftColumnWidth = 100;

    useEffect(() => {
        Axios.get(`${initialPage}&list=${list.toString()}`)
            .then((response: any) => {
                const result = response.data.data;
                setData(result);
                setTableHead(result.map((el: any) => el.country_name));
                setWidthArr(result.map(() => 100));

                const rowsHead = [
                    { label: 'Performance', value: 'performance_oriented' },
                    { label: 'Autocratic', value: 'autocratic' },
                    { label: 'Modesty', value: 'modesty' },
                    { label: 'Charisma', value: 'charisma' },
                    { label: 'Decisive', value: 'decisive' }];

                const recordData = [];
                for (let i = 0; i < rowsHead.length; i += 1) {
                    const rowData: any = [];
                    rowData.push(rowsHead[i].label);
                    recordData.push(rowData);
                }
                setRecData(recordData);

                const tableData = [];
                for (let i = 0; i < rowsHead.length; i += 1) {
                    const rowData = [];
                    for (let j = 0; j < result.length; j += 1) {
                        rowData.push(result[j][rowsHead[i].value]);
                    }
                    tableData.push(rowData);
                }
                setTabData(tableData);
            });

    }, []);

    return (
        <View
            style={{
                flex: 1,
                flexDirection: 'row',
                backgroundColor: '#eee',
                margin: 5
            }}
        >
            {/* Left Column */}
            <View
                style={{
                    width: leftColumnWidth,
                    backgroundColor: 'yellow',
                }}
            >
                {/* Blank Cell */}
                <View
                    style={{
                        height: headerHeight,
                        backgroundColor: primaryColor,
                        borderBottomWidth: 1,
                        borderBottomColor: borderColor,
                    }}
                ></View>
                {/* Left Container : scroll synced */}
                <ScrollView
                    ref={leftRef}
                    style={{
                        flex: 1,
                        backgroundColor: 'white',
                    }}
                    scrollEnabled={false}
                    showsVerticalScrollIndicator={false}
                >
                    <Table
                        borderStyle={{
                            borderWidth: 1,
                            borderColor,
                        }}
                    >
                        {recData.map((rowData: any, index: any) => (
                            <Row
                                key={index}
                                data={rowData}
                                height={headerHeight}
                                widthArr={[leftColumnWidth]}
                                style={styles.row}
                                textStyle={styles.textBold}
                            />
                        ))}
                    </Table>
                </ScrollView>
            </View>
            {/* Right Column */}
            <View
                style={{
                    flex: 1,
                    backgroundColor: 'white',
                }}
            >
                <ScrollView horizontal={true} bounces={false}>
                    <View>
                        <Table borderStyle={{ borderWidth: 1, borderColor }}>
                            <Row
                                data={tableHead}
                                widthArr={widthArr}
                                style={styles.head}
                                textStyle={{ ...styles.text, color: 'white' }}
                            />
                        </Table>
                        <ScrollView
                            ref={rightRef}
                            style={styles.dataWrapper}
                            scrollEventThrottle={16}
                            bounces={false}
                            onScroll={(e) => {
                                const { y } = e.nativeEvent.contentOffset;
                                leftRef.current?.scrollTo({ y, animated: false });
                            }}
                        >
                            <Table borderStyle={{ borderWidth: 1, borderColor }}>
                                {tabData.map((rowData: any, index: any) => (
                                    <Row
                                        key={index}
                                        data={rowData}
                                        height={headerHeight}
                                        widthArr={widthArr}
                                        style={styles.row}
                                        textStyle={styles.text}
                                    />
                                ))}
                            </Table>
                        </ScrollView>
                    </View>
                </ScrollView>
            </View>
        </View>
    )
};

const styles = StyleSheet.create({
    container: { flex: 1, padding: 16, paddingTop: 30, backgroundColor: '#eee' },
    head: { height: 40, backgroundColor: primaryColor },
    wrapper: { flexDirection: 'row' },
    title: { flex: 1, backgroundColor: '#f6f8fa' },
    row: { height: 'auto' },
    text: { textAlign: 'center' },
    textBold: { textAlign: 'center', fontWeight: 'bold' },
    dataWrapper: { marginTop: -1 },
});

export default CountryTable;