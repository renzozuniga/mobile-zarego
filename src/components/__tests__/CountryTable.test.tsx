import React from 'react'
import TestRenderer from 'react-test-renderer';
import CountryTable from '../CountryTable';

jest.useFakeTimers()

describe('src/components/CountryTable', () => {
    const mockedParams = {
        route: { params: { list: [] } }
    };

    const mockedParamsFull = {
        route: { params: { list: [1, 3, 5, 6, 8] } }
    };

    it('renders correctly', () => {
        const instance = TestRenderer.create(<CountryTable {...mockedParams} />).toJSON();
        expect(instance).toMatchSnapshot();
    })

    it('renders sections table', () => {
        const instance = TestRenderer.create(<CountryTable {...mockedParams} />).toJSON();
        expect(instance.children.length).toBe(2);
    })

    it('renders sections data', () => {
        const instance = TestRenderer.create(<CountryTable {...mockedParamsFull} />).toJSON();
        expect(instance.children[0].children.length).toBe(2);
    })
})