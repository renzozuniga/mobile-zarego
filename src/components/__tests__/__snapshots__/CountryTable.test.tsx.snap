// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/CountryTable renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#eee",
      "flex": 1,
      "flexDirection": "row",
      "margin": 5,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "yellow",
        "width": 100,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "black",
          "borderBottomColor": "#C1C0B9",
          "borderBottomWidth": 1,
          "height": 40,
        }
      }
    />
    <RCTScrollView
      scrollEnabled={false}
      showsVerticalScrollIndicator={false}
      style={
        {
          "backgroundColor": "white",
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            [
              undefined,
              {
                "borderBottomWidth": 1,
                "borderColor": "#C1C0B9",
                "borderLeftWidth": 1,
              },
            ]
          }
        />
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "backgroundColor": "white",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      bounces={false}
      horizontal={true}
    >
      <View>
        <View>
          <View
            style={
              [
                undefined,
                {
                  "borderBottomWidth": 1,
                  "borderColor": "#C1C0B9",
                  "borderLeftWidth": 1,
                },
              ]
            }
          >
            <View
              style={
                [
                  undefined,
                  {
                    "width": 1090,
                  },
                  {
                    "flexDirection": "row",
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "black",
                    "height": 40,
                  },
                ]
              }
            />
          </View>
          <RCTScrollView
            bounces={false}
            onScroll={[Function]}
            scrollEventThrottle={16}
            style={
              {
                "marginTop": -1,
              }
            }
          >
            <View>
              <View
                style={
                  [
                    undefined,
                    {
                      "borderBottomWidth": 1,
                      "borderColor": "#C1C0B9",
                      "borderLeftWidth": 1,
                    },
                  ]
                }
              />
            </View>
          </RCTScrollView>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
